### DISCUSSION TITLE
Custom validation with custom error message and invalid field

### DISCUSSION THREAD

**Discussion by chrismlusk** [verbb/formie#1796] [Q&A]
I'm adding custom validation to a form, and while I have the basics pretty much down, I'm struggling to get the desired errors shown at the end.

When a user submits a form, I need to hit a controller action to check if the zip code is in the client's service area. If not, I want to trigger a form error. The form error alert should explain why  the submission failed, so I want to include a custom message. I also want the form's zip code field to be marked as invalid and to show a field error.

In my own JS class, I started out with this:

```js
this.FormieFormBase = window.Formie.getFormById(this.options.formId);
this.$form = this.FormieFormBase.$form;
this.submitHandler = null;

this.$form.addEventListener('onFormieValidate', async (e) => {
  e.preventDefault();

  if (!(await this.isZipInServiceArea())) {
    this.submitHandler.formSubmitError();
  } else {
    this.submitHandler.submitForm();
  }
});
```

However, the `formSubmitError()` shows the form's default error message. After digging in the source code, I found some relevant methods on the Form Theme class, so I modified my code:

```js
if (!(await this.isZipInServiceArea())) {
  e.detail.invalid = true;
  this.FormieFormBase.formTheme.showFormAlert(MY_MESSAGE, 'error');
  this.FormieFormBase.formTheme.removeLoading();
  this.FormieFormBase.formTheme.scrollToForm();
}
```

This got me closer to what I want, but I'm still not sure what to do to get the target field marked as invalid. Any help?

**Comment by engram-design** [comment]
So you can go about adding validation this way, or you could use the [registerFormieValidation](https://verbb.io/craft-plugins/formie/docs/developers/javascript-api#the-registerFormieValidation-event) event, which is a little easier. Here you define your logic for validation, and the message to show in the field.

I'd probably encourage this approach, and despite thing changing a little in Formie 3 for Craft 5, it's an organised way to register validators.

Otherwise, to get the error to show up, you're a little bit on your own. Although using:

```
this.FormieFormBase.formTheme.validator.showError($input, { serverMessage: 'this is an error' });
```

is what we use to show server-side errors in the client-side validator library. Just supply it with the `$input` `<input>` element you want to target for the error.

Indeed, you can hook into the `showFormAlert()` function to change the main alert area at the top of the form, if you want to override that. By default, that's just a static message set in the form settings, and not part of the validators.

**Reply by chrismlusk** [reply]
Perfect! I am using the `registerFormieValidation` elsewhere, but I assumed since I need to call a controller action and need the validation for this specific field to be asynchronous (unlike the rest of the field validations), I'd need to handle it myself.

Thanks for the help!
